src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemetryContent.cs

```diff
@@ -18,49 +20,81 @@ public HttpTelemetryContent(EventBase evt)
                 evt.TryGetValue(MsalTelemetryBlobEventNames.ApiErrorCodeConstStrKey, out string errorCode);
                 evt.TryGetValue(MsalTelemetryBlobEventNames.ForceRefreshId, out string forceRefresh);
 
-                ApiId = apiId ?? string.Empty;
-                CorrelationId = correlationId ?? string.Empty;
-                LastErrorCode = errorCode ?? string.Empty;
+                ApiId.Add(apiId ?? string.Empty);
+                CorrelationId.Add(correlationId ?? string.Empty);
+                LastErrorCode.Add(errorCode ?? string.Empty);
                 ForceRefresh = forceRefresh ?? string.Empty;
             }
         }
 
-        public string LastErrorCode { get; set; } = string.Empty;
-        public string ApiId { get; set; } = string.Empty;
-        public string CorrelationId { get; set; } = string.Empty;
+        public /*For testing*/ HttpTelemetryContent(bool shouldClearCache)
+        {
+            if (shouldClearCache)
+            {
+                ApiId?.Clear();
+                CorrelationId?.Clear();
+                LastErrorCode?.Clear();
+            }
+        }
+
         public string ForceRefresh { get; set; } = string.Empty;
+        public List<string> ApiId { get; set; } = new List<string>();
+        public List<string> CorrelationId { get; set; } = new List<string>();
+        public List<string> LastErrorCode { get; set; } = new List<string>();
 
-        public string GetCsvAsPrevious()
+        public string GetCsvAsPrevious(int successfulSilentCallCount)
```

Comment:
> public string GetCsvAsPrevious(int successfulSilentCallCount) [](start = 7, length = 62)

Should collect 4k data and send that. The remaining data should stay. not sure why you clear things

Reply:
It's fairly complex and time consuming to keep creating a telemetry string and to measure it to see it is not > 4k. 4k is a LOT of data, so we're talking about a very rare corner case. IMO a proxy such as a number of api / http events is a better match for this.

Reply:
Fair point. However the time this is needed is when things goes wrong and it has been a long time since the last successful connection. I think nothing should prevent a proxy approach to this.

---
In reply to: [366462971](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1564#discussion_r366462971) [](ancestors = 366462971)

Reply:
The spec shows that the telemetry data can have multiple errors in it, right? In that sense, the length is not super consistent. What happens if we send an oversized request, does it return error? Then we lose the data either way. Can we really truncate it though (the csv will be malformed?)? We could consider using StringBuilder and/or Length property, maybe it's more performant.

Reply:
Completely agree, cutting the data will result in malformed CSV. Might as well not send it :). Not sure about Lenght property on StringBuilder. In C# strings are UTF-8, so size of char is 1 or 2 bytes (see https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/). Might as well consider it 2 bytes or transform to byte array and measure that. 

During telemetry meeting there was discussions about aggregating this data. Not sure if any proposal were sent out on this?